File: frontend/src/services/deviceConfig.ts
import {
  GetDeviceConfigResponseDuplexEnum,
  SetDeviceConfigRequest,
} from "../generated/api";
import { api } from "../backendApi";

export interface DeviceConfigState {
  ipAddress: string;
  netMask: string;
  bitRate: number;
  duplex: GetDeviceConfigResponseDuplexEnum;
}

export async function getDeviceConfig(): Promise<DeviceConfigState> {
  const getDeviceConfigResponse = await api.getDeviceConfig();
  const deviceConfigState = {
    ipAddress: getDeviceConfigResponse.data.ipAddress,
    netMask: getDeviceConfigResponse.data.netMask,
    bitRate: getDeviceConfigResponse.data.bitRate,
    duplex: getDeviceConfigResponse.data.duplex,
  } as DeviceConfigState;

  return deviceConfigState;
}

export async function setDeviceConfig(newConfig: DeviceConfigState) {
  const setDeviceConfigRequest = {
    ipAddress: newConfig.ipAddress,
    netMask: newConfig.netMask,
    bitRate: newConfig.bitRate,
    duplex: newConfig.duplex,
  } as SetDeviceConfigRequest;
  await api.setDeviceConfig(setDeviceConfigRequest);
}

File: device-api/rest-api/device_config.go
package rest_api

import (
	"context"
	"encoding/json"
	"io/ioutil"
	"log"
	"net/http"

	openapi "github.com/device-admin-org/device-admin/device-api/generated/go"
)

var deviceConfigFilePath string = "/mnt/shared_data/device_config.json"

func GetDeviceConfig(ctx context.Context) (openapi.ImplResponse, error) {
	log.Printf("GetDeviceConfig called\n")
	data, err := ioutil.ReadFile(deviceConfigFilePath)
	if err != nil {
		log.Printf("Could not read json file, using defaults: %s\n", err)
		response := openapi.GetDeviceConfigResponse{
			IpAddress: "1.2.3.4",
			NetMask:   "255.255.255.0",
			BitRate:   115200,
			Duplex:    "full",
		}
		return openapi.Response(http.StatusOK, response), nil
	}
	var response openapi.GetDeviceConfigResponse
	err = json.Unmarshal(data, &response)
	if err != nil {
		log.Printf("Could not unmarshal json: %s\n", err)
		return openapi.Response(http.StatusOK, nil), nil
	}
	log.Printf("Unmarshalled json file: %v\n", response)
	return openapi.Response(http.StatusOK, response), nil
}

func SetDeviceConfig(ctx context.Context, request openapi.SetDeviceConfigRequest) (openapi.ImplResponse, error) {
	log.Printf("SetDeviceConfig called with request: %v\n", request)
	jsonData, err := json.Marshal(request)
	if err != nil {
		log.Printf("Could not marshal json: %s\n", err)
		return openapi.Response(http.StatusOK, nil), nil
	}
	log.Printf("Marshalled json: %s\n", &jsonData)
	err = ioutil.WriteFile(deviceConfigFilePath, jsonData, 0)
	if err != nil {
		log.Printf("Could not write marshalled json file: %s\n", err)
		return openapi.Response(http.StatusOK, nil), nil
	}
	return openapi.Response(http.StatusOK, nil), nil
}
